crates/aptos-jwk-consensus/src/jwk_manager/mod.rs

```diff
@@ -0,0 +1,408 @@
+                JWKObserver::spawn(
+                    self.my_addr,
+                    provider.name.clone(),
+                    provider.config_url.clone(),
+                    Duration::from_secs(10),
+                    local_observation_tx.clone(),
+                )
+            })
+            .collect();
+
+        let mut close_rx = close_rx.into_stream();
+
+        while !self.stopped {
+            let handle_result = tokio::select! {
+                jwk_updated = jwk_updated_rx.select_next_some() => {
+                    let ObservedJWKsUpdated { jwks, .. } = jwk_updated;
+                    self.reset_with_on_chain_state(jwks)
+                },
+                (_sender, msg) = rpc_req_rx.select_next_some() => {
+                    self.process_peer_request(msg)
+                },
+                qc_update = qc_update_rx.select_next_some() => {
+                    self.process_quorum_certified_update(qc_update)
+                },
+                (issuer, jwks) = local_observation_rx.select_next_some() => {
+                    let jwks = jwks.into_iter().map(JWKMoveStruct::from).collect();
+                    self.process_new_observation(issuer, jwks)
+                },
+                ack_tx = close_rx.select_next_some() => {
+                    self.tear_down(ack_tx.ok()).await
+                }
+            };
+
+            if let Err(e) = handle_result {
+                error!("[JWK] handling_err={}", e);
+            }
+        }
+    }
+
+    async fn tear_down(&mut self, ack_tx: Option<oneshot::Sender<()>>) -> Result<()> {
+        self.stopped = true;
+        let futures = std::mem::take(&mut self.jwk_observers)
+            .into_iter()
+            .map(JWKObserver::shutdown)
+            .collect::<Vec<_>>();
+        join_all(futures).await;
+        if let Some(tx) = ack_tx {
+            let _ = tx.send(());
+        }
+        Ok(())
+    }
+
+    /// Triggered by an observation thread periodically.
+    pub fn process_new_observation(
+        &mut self,
+        issuer: Issuer,
+        jwks: Vec<JWKMoveStruct>,
+    ) -> Result<()> {
+        let state = self.states_by_issuer.entry(issuer.clone()).or_default();
+        state.observed = Some(jwks.clone());
+        if state.observed.as_ref() != state.on_chain.as_ref().map(ProviderJWKs::jwks) {
+            let observed = ProviderJWKs {
+                issuer: issuer.clone(),
+                version: state.on_chain_version() + 1,
+                jwks,
+            };
+            let signature = self
+                .consensus_key
+                .sign(&observed)
+                .map_err(|e| anyhow!("crypto material error occurred duing signing: {}", e))?;
+            let abort_handle = self.certified_update_producer.start_produce(
+                self.epoch_state.clone(),
+                observed.clone(),
+                self.qc_update_tx.clone(),
+            );
+            state.consensus_state = ConsensusState::InProgress {
```

Comment:
It feels safer to me to update local state first and then do broadcast. Currently it is fine since you passed in observed. Maybe in the future broadcast depends on local state then it may cause some issue.

Reply:
it's even more safe if the order doesn't matter, no?